File: input-structures/src/smooth.c
#include <stdio.h>
#include <math.h>

//I need to access in range 0..Nmax-1
int smooth_arr_zm_fur(double *Fs, int Nmax, double smooth_intensity, double *Fi, double *Ftd){
  int n;
  double sgn, f1, f2;
//Should be noted: Fi[Nmax -1] doesn't initialized!
  for(n = 0; n<= Nmax - 2; n++)
    Fi[n] = smooth_intensity * (Fs[n+1] - Fs[n]);
    //So i need to determine somehow Fi[Nmax-1].
Fi[Nmax-1] = 0; //Added this!
  for(n = 1; n<= Nmax - 2; n++)
    Ftd[n] = Fs[n] + (Fi[n] - Fi[n-1]);

  Ftd[0] = Fs[0];
  Ftd[Nmax - 1] = Fs[Nmax - 1];

  for(n = 0; n < Nmax - 1; n++)
    Fs[n] = Ftd[n+1] - Ftd[n];
//Main calculation ++++++++++++++++++++++++++++++
  for(n = 0; n<= Nmax - 2; n++)
  {
    Fi[n] = smooth_intensity * (Ftd[n+1] - Ftd[n]);
    sgn = -1;
    if (Fi[n] >=0)
      {
        sgn = 1;
        Fi[n] = fabs(Fi[n]);
      }
      
    if (n==0){
      f2 = sgn * Fs[1];
      if( f2<Fi[0] )
        Fi[0] = f2;
    }
    else if (n == Nmax - 2)
    {
      f1 = sgn * Fs[Nmax - 2];
      if (f1 < Fi[Nmax - 1])
      Fi[Nmax - 1] = f1;
    }
  //This means to use if n!= 0 and end_index of vector(Nmax-1)
    else
    {
      f1 = sgn * Fs[n-1];
      f2 = sgn * Fs[n+1];
      if (f1 < Fi[n])
      Fi[n] = f1;
      if (f2 < Fi[n])
      Fi[n] = f2;
    }
if (Fi[n] < 0)
      Fi[n] = 0;
      else
      Fi[n] = sgn * Fi[n];
}
//---------------------------------------
for(n = 1; n<= Nmax - 1; n++){
      Fs[n] = Ftd[n] - (Fi[n] - Fi[n-1]);
      Fs[0] = Ftd[0];
  }
      return 0;
}
void callback(){

}
int add(int a, int b) {
    return a + b;
}

File: input-structures/src/smooth.rs
pub fn smooth_zf_rs(fs: &mut Vec<f64>, nmax: usize, smooth_int: f64, fi:  &mut Vec<f64>, ftd:  &mut Vec<f64>){
    let mut sign: i64 = -1;
    let mut f1: f64;
    let mut f2: f64;

    for n in 0.. nmax-1{
        fi[n]= smooth_int * (fs[n+1] - fs[n]);
    }
//So i need to determine somehow Fi[nmax-1].
    fi[nmax-1] = 0.0; //Notice, please.
    for n in 1.. nmax-1{
        ftd[n] = fs[n] + (fi[n] - fi[n-1]);
    }
    ftd[0] = fs[0];
    ftd[nmax -1] = fs[nmax -1];
    for n in 0 .. (nmax - 1){
        fs[n] = ftd[n+1] - ftd[n];
    }
//Main calculations ++++++++++++++++++++++++++++++++++++++++
    for n in 0..nmax-1{
        fi[n] = smooth_int * (ftd[n+1] - ftd[n]);
        let sgn = fi[n].signum() as i64; // copy sign from calculated value
        if sgn >= 0 {
            fi[n] = fi[n].abs();
        sign = sgn
        }
        if n==0{
            f2= sign as f64 * fs[1];
            if f2< fi[0]{
                fi[0] = f2;
            }
        }
    else if n== nmax-2{//Should be noticed!
        f1 = sign as f64 * fs[nmax - 2];
        if  f1 < fi[nmax - 1]//What value??? - 0;
        {
            fi[nmax - 1] = f1;
        }
    }
    else
    {
        f1 = sign as f64 * fs[n-1];
        f2 = sign as f64 * fs[n+1];
        if f1 < fi[n]{
            fi[n] = f1;
        }
        if f2 < fi[n]{
            fi[n] = f2;
        }
    }
    if fi[n] < 0.0
        {fi[n] = 0.0;}
    else
      {fi[n] = sign as f64 * fi[n];}
    }
//-------------------------------------------
    for n in 1..nmax{
        fs[n] = ftd[n] - (fi[n] - fi[n-1]);//fi[last] = 0;
        fs[0] = ftd[0];
    }
}
